editoast/src/main.rs

```diff
@@ -134,6 +141,70 @@ fn log_received_request(req: &ServiceRequest) {
     log::info!(target: "actix_logger", "{} RECEIVED", request_line);
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+struct Claims {
+    sub: String,
+    exp: usize,
+}
+
+async fn get_jwks(
+    oidc_jwk_uris: Option<String>,
+    req: &ServiceRequest,
+) -> Result<JwkSet, reqwest::Error> {
+    let cache_mutex = req.app_data::<Data<Mutex<JwkSet>>>().unwrap();
+    let mut cache = cache_mutex.lock().unwrap();
+    if cache.keys.is_empty() {
+        let mut jwks = JwkSet { keys: vec![] };
+        let oidc_jwk_uris: Vec<String> = oidc_jwk_uris
+            .expect("OIDC_JWK_URIS must be set")
+            .split(',')
+            .map(|s| s.trim().to_string()) // trim to remove any unwanted spaces
+            .collect();
+        println!("{:?}", oidc_jwk_uris);
+        for uri in &oidc_jwk_uris {
+            let body = reqwest::get(uri).await?.text().await?;
+            let mut response: JwkSet = serde_json::from_str(&body).unwrap();
+            jwks.keys.append(&mut response.keys)
+        }
+        cache.keys = jwks.keys.clone();
+    }
+    Ok(cache.clone())
+}
+
+async fn validator(
+    oidc_jwk_uris: Option<String>,
+    req: ServiceRequest,
+    credentials: BearerAuth,
+) -> Result<ServiceRequest, (ActixError, ServiceRequest)> {
+    let jwt_header = decode_header(&credentials.token()).unwrap();
+    let token_kid = jwt_header.kid.ok_or("No kid in token").unwrap();
+    let jwks = get_jwks(oidc_jwk_uris, &req)
+        .await
+        .unwrap_or(JwkSet { keys: vec![] });
+    let jwk_to_use = jwks.find(&token_kid);
+    let jwk = match jwk_to_use {
```

Comment:
If the `Jwk` is not found in the set we should update the set and try again. If we don't find it again then we can return an error.